Rewrite of Concepts and Architecture (#239)
pinot-contrib/pinot-docs: basics/architecture.md

---
description: >-

>   Understand how the components of Apache Pinot™ work together to create a scalable OLAP database that can delivery low-latency, high-concurrecy queries at scale.

Reviewer: ```suggestion
  Understand how the components of Apache Pinot™ work together to create a scalable OLAP database that can deliver low-latency, high-concurrency queries at scale.
```

---

# Architecture

{% hint style="info" %}
We recommend that you read [Basic Concepts](concepts.md) to better understand the terms used in this guide.
{% endhint %}

> Apache Pinot™ is a distributed OLAP database designed to serve real-time, user-facing use cases, which means handling large volumes of data and many concurrent queries with very low query latencies. This imposes some unique and demanding requirements:

Reviewer: ```suggestion
Apache Pinot™ is a distributed OLAP database designed to serve real-time, user-facing use cases, which means handling large volumes of data and many concurrent queries with very low query latencies. Pinot supports the following requirements:
```

> Apache Pinot™ is a distributed OLAP database designed to serve real-time, user-facing use cases, which means handling large volumes of data with very low query latencies in the face of many concurrent queries. This imposes some unique and demanding requirements:

Reviewer: ```suggestion
Apache Pinot™ is a distributed OLAP database designed to serve real-time, user-facing use cases, which means handling large volumes of data and many concurrent queries with very low query latencies. This imposes some unique and demanding requirements:
```

* Ultra low-latency queries (as low as 10ms P95)
* High query concurrency (as many as 100,000 queries per second)
* High data freshness (streaming data available for query immediately upon ingestion)
* Large data volume (up to petabytes)

## Distributed design principles

> To accommodate large data volumes with stringent latency and concurrency requirements, Pinot is designed as a distributed database. As a distributed system, it has these goals:

Reviewer: ```suggestion
To accommodate large data volumes with stringent latency and concurrency requirements, Pinot is designed as a distributed database that supports the following requirements:
```

> 

Reviewer: ```suggestion
```

> * **Highly available**: Pinot has no single point of failure. When operators choose data replication, the cluster can continue to serve queries when a node goes down.

Reviewer: ```suggestion
* **Highly available**: Pinot has no single point of failure. When tables are configured for replication, and a node goes down, the cluster is able to continue processing queries.
```

> * **Highly available**: Pinot has no single point of failure. When operators choose data replication, the cluster can continue to serve queries when a node goes down.

Reviewer: does "operators" refer to running services or a user who specifies the data replication?

Author: It refers to people. :)

Reviewer: hmmm, so would this make sense:

- When a cluster is configured to replicate data, and a node goes down, the cluster is able to continue processing queries.

Author: Not quite. Replication is configured at the table level, so it would have to be:

> When tables are configured for replication, and a node goes down, the cluster is able to continue processing queries.

I was trying to avoid the passive voice, but this gets the mysterious operators out of the picture. :)

> * **Horizontally scalable**: Operators can scale a Pinot cluster by adding new nodes when the workload increases. There are even two node types to scale query volume, query complexity, and data size independently.

Reviewer: Re: _There are even two node types..._
can we clarify the two?

Author: Yes, I'll push that change.

> * **Immutable data**: Pinot treats all stored data as if it is immutable, which gives designers the ability to make simplifying assumptions elsewhere in the system. However, Pinot still supports upserts on streaming entity data and background purges of data to comply with data privacy regulations.

Reviewer: what's the difference between operators and designers?

Reviewer: if we drop "entity" does that change the meaning?

Author: I think "entity" is important, yes. It distinguishes between streaming events and entity changelogs. This is too much to explain inline, would distract from the point, and the sentence survives if the distinction is unclear, but if the reader does understand the distinction, it's quite meaningful.

Author: I should have said, "which gives _its_ designers the ability...." I'll push that change. Does that fix things?

Reviewer: not sure if accurate, but reading this like:

Pinot treats all stored data as if it is immutable, which makes system design and maintenance easier.

Author: What I'm really going for is this idea:

> Pinot assumes all stored data is immutable, which helps simplify the parts of the system that handle data storage and replication.

Maybe just that instead?

Reviewer: to be sure, does it simplify the parts of the system, or simplify the handling? or is that the same thing?

> * **Immutable data**: Pinot treats all stored data as if it is immutable, which gives designers the ability to make simplifying assumptions elsewhere in the system. However, Pinot still supports upserts on streaming entity data and background purges of data to comply with data privacy regulations.

Reviewer: ```suggestion
* **Immutable data**: Pinot assumes all stored data is immutable, which helps simplify the parts of the system that handle data storage and replication. However, Pinot still supports upserts on streaming entity data and background purges of data to comply with data privacy regulations.
```

* **Dynamic configuration changes**: Operations like adding new tables, expanding a cluster, ingesting data, modifying an existing table, and adding indexes do not impact query availability or performance.

## Core components

As described in [Apache Pinot™ Concepts](concepts.md), Pinot has four node types:

* [Controller](components/cluster/controller.md)
* [Broker](components/cluster/broker.md)
* [Server](components/cluster/server.md)
* [Minion](components/cluster/minion.md)

![](<../.gitbook/assets/Pinot-architecture (1).svg>)

> ### Apache Helix and Zookeeper

Reviewer: ```suggestion
### Apache Helix and ZooKeeper
```

> Distributed systems do not maintain themselves, and in fact require sophisticated scheduling and resource management to function. Pinot uses [Apache Helix](http://helix.apache.org/) for this purpose. Helix exists as an independent project, but it was designed by the original creators of Pinot for Pinot's own cluster management purposes, so the architectures of the two systems are well-aligned. Helix takes the form of a process on the controller, plus embedded agents on the brokers and servers. It uses [Apache Zookeeper](https://zookeeper.apache.org/) as a fault-tolerant, strongly consistent, durable state store.

Reviewer: ```suggestion
Distributed systems do not maintain themselves, and require sophisticated scheduling and resource management to function. Pinot uses [Apache Helix](http://helix.apache.org/) for this purpose. Helix exists as an independent project, but it was designed by the original creators of Pinot for Pinot's own cluster management purposes, so the architectures of the two systems are well-aligned. Helix takes the form of a process on the controller, plus embedded agents on the brokers and servers. It uses [Apache Zookeeper](https://zookeeper.apache.org/) as a fault-tolerant, strongly consistent, durable state store.
```

> Distributed systems do not maintain themselves, and in fact require sophisticated scheduling and resource management to function. Pinot uses [Apache Helix](http://helix.apache.org/) for this purpose. Helix exists as an independent project, but it was designed by the original creators of Pinot for Pinot's own cluster management purposes, so the architectures of the two systems are well-aligned. Helix takes the form of a process on the controller, plus embedded agents on the brokers and servers. It uses [Apache Zookeeper](https://zookeeper.apache.org/) as a fault-tolerant, strongly consistent, durable state store.

Reviewer: ```suggestion
Distributed systems do not maintain themselves, and in fact require sophisticated scheduling and resource management to function. Pinot uses [Apache Helix](http://helix.apache.org/) for this purpose. Helix exists as an independent project, but it was designed by the original creators of Pinot for Pinot's own cluster management purposes, so the architectures of the two systems are well-aligned. Helix takes the form of a process on the controller, plus embedded agents on the brokers and servers. It uses [Apache ZooKeeper](https://zookeeper.apache.org/) as a fault-tolerant, strongly consistent, durable state store.
```

* **Spectator**: These nodes _see_ things, observing the evolving state of the _participants_ through events pushed to the spectator. Because Pinot brokers need to know which servers host which segments, they are spectators.
* **Controller**: This node observes and manages the state of participant nodes. The controller is responsible for coordinating all state transitions in the cluster and ensures that state constraints are satisfied while maintaining cluster stability.

In addition, Helix defines two logical components to express its storage abstraction:

* **Partition**. A unit of data storage that lives on at least one participant. Partitions may be replicated across multiple participants. A Pinot segment is a partition.
* **Resource**. A logical collection of partitions, providing a single view over a potentially large set of data stored across a distributed system. A Pinot table is a resource.

In summary, the Pinot architecture maps onto Helix components as follows:

|Pinot Component|Helix Component|
|----------|-----------|
|Segment|**Helix Partition**|
|Table|**Helix Resource**|

> |Controller|Runs the Helix agent that drives the overall state of the cluster (also called the Helix Controller)|

Reviewer: ```suggestion
|Controller| **Helix Controller** or Helix agent that drives the overall state of the cluster|
```

|Server|**Helix Participant**|

> |Broker|A **Helix Spectator** that observes the cluster for changes in the state of segments and servers. In order to support multi-tenancy, brokers are also modeled as **Helix Participants**.|

Reviewer: ```suggestion
|Broker|A **Helix Spectator** that observes the cluster for changes in the state of segments and servers. To support multi-tenancy, brokers are also modeled as **Helix Participants**.|
```

|Minion|**Helix Participant** that performs computation rather than storing data|

> Helix uses Zookeeper to maintain cluster state. Helix spectators receive notifications directly from Zookeeper of changes in cluster state (which correspond to changes in ZNodes). Zookeeper stores the following information about the cluster:

Reviewer: ```suggestion
Helix uses ZooKeeper to maintain cluster state. ZooKeeper sends Helix spectators notifications of changes in cluster state (which correspond to changes in ZNodes). Zookeeper stores the following information about the cluster:
```

|Resource|Stored Properties|
|--------|-----------------|
|Controller|<ul><li>Controller that is assigned as the current leader</li></ul>|
|Servers and Brokers|<ul><li>List of servers and brokers</li><li>Configuration of all current servers and brokers</li><li>Health status of all current servers and brokers</li></ul>|

> |Tables|<ul><li>List of tables</li><li>Table configurations</li><li>Table schema</li><li>List of the table's segments</li></ul>|

Reviewer: can there be multiple table schemas?

|Segment|<ul><li>Exact server locations of a segment</li><li>State of each segment (online/offline/error/consuming)</li><li>Metadata about each segment</li></ul>|

> Zookeeper is a first-class citizen of a Pinot cluster, so it is appropriate to use the well-known `ZNode` structure for operations and troubleshooting purposes. Be advised that this structure can change in future Pinot releases.

Reviewer: ```suggestion
Zookeeper, as a first-class citizen of a Pinot cluster, may use the well-known `ZNode` structure for operations and troubleshooting purposes. Be advised that this structure can change in future Pinot releases.
```

#### Controller REST interface

The controller provides a REST interface that allows read and write access to all logical storage resources (e.g., servers, brokers, tables, and segments). See [Pinot Data Explorer](components/exploring-pinot.md) for more information on the web-based admin tool.

### Broker

The [broker's](components/cluster/broker.md) responsibility is to route queries to the appropriate [server](components/cluster/server.md) instances, or in the case of multi-stage queries, to compute a complete query plan and distribute it to the servers required to execute it. The broker collects and merges the responses from all servers into a final result, then sends the result back to the requesting client. The broker exposes an HTTP endpoint that accepts SQL queries in JSON format and returns the response in JSON.

> Each broker maintains a query routing table. The routing table maps segments to the servers that store them. (When replication is configured on a table, each segment is stored on more than one server.) The broker computes multiple routing tables depending on the configured [routing](operators/tuning/routing.md) strategy for a table. The default strategy is to balance the query load across all available servers.

Reviewer: ```suggestion
Each broker maintains a query routing table. The routing table maps segments to the servers that store them. (When replication is configured on a table, each segment is stored on more than one server.) The broker computes multiple routing tables depending on the configured [routing](operators/operating-pinot/tuning/routing.md) strategy for a table. The default strategy is to balance the query load across all available servers.
```

}
```

#### Query processing

> Every query processed by a broker uses the single-stage engine or the [multi-stage engine](https://docs.pinot.apache.org/reference/multi-stage-engine). For single stage queries, the broker does the following:

Reviewer: ```suggestion
Every query processed by a broker uses the single-stage engine or the [multi-stage engine](https://docs.pinot.apache.org/reference/multi-stage-engine). For single-stage queries, the broker does the following:
```

* Computes query routes based on the routing strategy defined in the [table](components/table/) configuration.

> * Computes the list of segments to query on each [server](components/cluster/server.md). (See [routing](operators/tuning/routing.md) for further details on this process.)

Reviewer: ```suggestion
* Computes the list of segments to query on each [server](components/cluster/server.md). (See [routing](operators/operating-pinot/tuning/routing.md) for further details on this process.)
```

Because offline tables tend to have long retention periods, offline servers tend to scale based on the size of the data they store.

![](<../.gitbook/assets/OfflineServer (4).jpg>)

#### Real-time servers

Real-time servers ingest data from streaming sources, like Apache Kafka®, Apache Pulsar®, or AWS Kinesis. Streaming data ends up in conventional segment files just like batch data, but is first accumulated in an in-memory data structure known as a consuming segment. Each message consumed from a streaming source is written immediately to the relevant consuming segment, and is available for query processing from the consuming segment immediately, since consuming segments participate  in query processing as first-class citizens. Consuming segments get flushed to disk periodically based on a completion threshold, which can be calculated by row count, ingestion time, or segment size. A flushed segment on a real-time table is called a _completed_ segment, and is functionally equivalent to a segment created during offline ingest.

Real-time servers tend to be scaled based on the rate at which they ingest streaming data.

![](../.gitbook/assets/real-time-flow.svg)

### Minion

> A Pinot [minion](components/cluster/minion.md) is an optional cluster component that executes background tasks on table data apart from the query processes performed by brokers and servers. Minions run on independent hardware resources, and are responsible for executing _minion tasks_ as directed by the controller. Examples of minon tasks include converting batch data from a standard format like Avro or JSON into segment files to be loaded into an offline table, and rewriting existing segment files to purge records as required by data privacy laws like GDPR. Minion tasks can run once or be scheduled to run periodically.

Reviewer: ```suggestion
A Pinot [minion](components/cluster/minion.md) is an optional cluster component that executes background tasks on table data apart from the query processes performed by brokers and servers. Minions run on independent hardware resources, and are responsible for executing _minion tasks_ as directed by the controller. Examples of minion tasks include converting batch data from a standard format like Avro or JSON into segment files to be loaded into an offline table, and rewriting existing segment files to purge records as required by data privacy laws like GDPR. Minion tasks can run once or be scheduled to run periodically.
```

}
```

For multi-stage queries, the broker performs the following:

* Computes a query plan that runs on multiple sets of servers. The servers selected for the first stage are selected based on the segments required to execute the query, which are determined in a process similar to single-stage queries.

> * Sends the relevant portions of the query plan to each stage of servers in the cluster. A stage may involve one or many servers.

Reviewer: ```suggestion
* Sends the relevant portions of the query plan to one or more servers in the cluster for each stage of the query plan. 
```

* The servers that received query plans each execute their part of the query. For more details on this process, read about the [multi-stage engine](https://docs.pinot.apache.org/reference/multi-stage-engine).
* The broker receives a complete result set from the final stage of the query, which is always a single server.
* The broker sends the query result to the client.

### Server

> [Servers](components/cluster/server.md) host [segments](components/table/segment/) on locally attached storage and perform the computational work of query processing on those segments. By convention, operators speak of "real-time" and "offline" servers, although there is no difference in the server process itself or even its configuration that distinguishes between the two. This is merely a convention reflected in the [table](components/table/) assignment strategy to confine the two different kinds of workloads to two groups of physical instances, since the performance-limiting factors differ between the two kinds of workloads. For example, offline servers might optimize for larger storage capacity, whereas real-time servers might optimize for memory and CPU cores.

Reviewer: ```suggestion
[Servers](components/cluster/server.md) host [segments](components/table/segment/) on locally attached storage and process queries on those segments. By convention, operators speak of "real-time" and "offline" servers, although there is no difference in the server process itself or even its configuration that distinguishes between the two. This is merely a convention reflected in the [table](components/table/) assignment strategy to confine the two different kinds of workloads to two groups of physical instances, since the performance-limiting factors differ between the two kinds of workloads. For example, offline servers might optimize for larger storage capacity, whereas real-time servers might optimize for memory and CPU cores.
```

#### Offline servers

> Offline servers host segments created by ingesting batch data. These segments are created apart from the hosting server itself and transferred to the server through a process coordinated by the controller according to the table's replication factor and segment assignment strategy. Typically, the controller directs a newly uploaded segment to be written to the [deep store](https://docs.pinot.apache.org/basics/components/table/segment/deep-store), then informs the affected servers to download the segment from there. It then informs the brokers that a new segment exists, after which it is available to participate in queries.

Reviewer: ```suggestion
Offline servers host segments created by ingesting batch data. The controller writes these segments to the offline server according to the table's replication factor and segment assignment strategy. 
Typically, the controller writes new segments to the [deep store](https://docs.pinot.apache.org/basics/components/table/segment/deep-store), and affected servers download the segment from deep store. The controller then notifies brokers that a new segment exists, and is available to participate in queries.
```

Real-time servers tend to be scaled based on the rate at which they ingest streaming data.

![](../.gitbook/assets/real-time-flow.svg)

### Minion

A Pinot [minion](components/cluster/minion.md) is an optional cluster component that executes background tasks on table data apart from the query processes performed by brokers and servers. Minions run on independent hardware resources, and are responsible for executing _minion tasks_ as directed by the controller. Examples of minon tasks include converting batch data from a standard format like Avro or JSON into segment files to be loaded into an offline table, and rewriting existing segment files to purge records as required by data privacy laws like GDPR. Minion tasks can run once or be scheduled to run periodically.

> Minions isolate the computational burden of out-of-band data processing from the servers. Although a Pinot cluster can function with without minions, they are typically present to support routine tasks like batch data ingest

Reviewer: ```suggestion
Minions isolate the computational burden of out-of-band data processing from the servers. Although a Pinot cluster can function without minions, they are typically present to support routine tasks like ingesting batch data.
```

## Data ingestion overview

Pinot [tables](components/table/) exist in two varieties: offline (or batch) and real-time. Offline tables contain data from batch sources like CSV, Avro, or Parquet files, and real-time tables contain data from streaming sources like like Apache Kafka®, Apache Pulsar®, or AWS Kinesis.

### Offline (batch) ingest

![](<../.gitbook/assets/OfflineServer (4).jpg>)

Pinot ingests batch data using an [ingestion job](data-import/batch-ingestion/), which follows a process like this:

1. The job transforms a raw data source (such as a CSV file) into [segments](components/table/segment/). This is a potentially complex process resulting in a file that is typically several hundred megabytes in size.

> 2. The job then transfers the file to the cluster's [deep store](components/table/segment/deep-store) and notifies the [controller](components/cluster/controller.md) that a new segment exists.

Reviewer: ```suggestion
2. The job then transfers the file to the cluster's [deep store](components/table/segment/deep-store.md) and notifies the [controller](components/cluster/controller.md) that a new segment exists.
```
